### ISSUE TITLE
Atomic CPU operations for parallel addition, subtraction, multiplication and division.

### ISSUE THREAD

**Issue by v0dro** [pytorch/pytorch#33065]
## 🚀 Feature

Portable parallel atomic aritmetic operations like addition, subtraction, division and multiplication
for CPU.

## Motivation

Scatter reduction operations came up in issue https://github.com/pytorch/pytorch/issues/22378 . Performing
these operations safely in parallel requires the use of atomic operations for performing various
operations like addition, subtraction, division and multiplication. @nikitaved started porting
the `scatter_add` from TH to ATen in https://github.com/pytorch/pytorch/pull/31662. In this issue,
he implements a new kernel using `TensorIterator` called [cpu\_scatter\_gather\_base\_kernel](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L92),
which currently cannot run in parallel since atomic operations in parallel would be buggy.

The issue raised by @ngimel regarding documentation for the scatter and scatter add methods (https://github.com/pytorch/pytorch/issues/31776)
raises some interesting issues regarding use of non-unique indices for scatter operations.
If these operations are to made parallel, it very important that operations involving the
same index are atomic so that multiple threads accessing the same space in memory perform
their operations correctly.

## Pitch

Currently the `scatter_add` operation on CPU is serial due to lack of a `atomicAdd` operation
similar to what exists in the [CUDA backend](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCAtomics.cuh). Implementing this operation on CPU
should open up possibilities for paralellization of scatter reductions (not restricted to addition).

## Additional context

Eventually, we want to implement deterministic scatter operations in pytorch (https://github.com/pytorch/pytorch/issues/23151),
and performing these in parallel will of course require the use of atomic operations.

Additionally, @nikitaved has implemented fast gather operations (https://github.com/pytorch/pytorch/pull/32425), and he says that since the
`scatter_add` is in the backward of gather, and gather is parallel, having a `scatter_add`
that is parallel is an important need.

cc @VitalyFedyunin @ngimel

**Comment by fmassa** [comment]
One point you mentioned that I'd like to understand a bit better:

> Eventually, we want to implement deterministic scatter operations in pytorch (#23151),
and performing these in parallel will of course require the use of atomic operations.

I thought that using atomic operations didn't guarantee determinism, but instead the order of the operations being the same was necessary. Isn't it the case?

**Comment by v0dro** [comment]
One of the causes for the loss of determinism is non-ordering of the operations, hence this is step in that direction. The other cause if of course algorithm-specific and will be explored later.

**Comment by ngimel** [comment]
In case of scatter and other indexing operations, atomics are necessary for correctness, but they don't guarantee determinism (in fact, they guarantee non-deterministic results for all practical purposes, if one also does not somehow ensure that atomics operations are performed in the same order). Atomic operations for scatter-like ops are necessary to enable parallel execution, so in that sense this issue is important. 
If one is to deterministically implement scatterOp variants, different approaches are required.

**Comment by v0dro** [comment]
So I tried implementing atomic add with `std::atomic` but I find that it only works with data copying, and its not feasible to copy data back and forth from an atomic object (slow). I tried to make it work with pointers but `compare_exchange_weak` does not allow writing to the value of a pointer.

Therefore, we are left with two possible approaches:
1. Use a simple mutex and possible incur thread contention overheads.
2. Use compiler-specific atomic built-in operations like GCC's `__atomic` and have separate implementations for each supported compiler.

I'll try and benchmark both and see which is faster for which cases. However the code maintainability will of course decrease for `2`.
